### PR TITLE
Use GCP Datastore to store comments, add pagination

### PR DESCRIPTION
**Pull request by n-wach**
Waiting on #17

### CODE REVIEW

**Comment by n-wach** [comment]
Not sure why there are so many changes under "Files Changed"...  Just look at 03ba865 for the changes relevant to this PR

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  private final Date created;
+
+  public Comment(String content, Date created) {
+    this.content = content;
+    this.created = created;
+  }
+
+  public Comment(String content) {
+    this(content, new Date());
+  }
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:32]: When would we not need a date when a Comment is created. Perhaps we should force it?

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import java.util.Date;
+
+/** Class describing a comment */
+public final class Comment {
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:20]: This class looks like a data class, I would suggest @AutoValue for this and user a builder.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import java.util.Date;
+
+/** Class describing a comment */
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:19]: nit: /** Contains data related to a comment*/ ?

Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:19]:
```suggestion
/** Contains data related to a comment*/ 
```
WDYT?

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  public Comment(String content, Date created) {
+    this.content = content;
+    this.created = created;
+  }
+
+  public Comment(String content) {
+    this(content, new Date());
+  }
+
+  public String getContent() {
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:34]: nit:It's good practice to have comments for all public non-overriden methods.

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+package com.google.sps.data;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import com.google.sps.data.CommentThread;
+
+/** Class describing a series of comments, with a cursor for pagination */
+public final class CommentThread {
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/data/CommentThread.java:23]: I'll suggest an ```@AutoValue``` for this as well, it should also force the list to be Immutable IIRC.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    // If this servlet is passed a cursor parameter, let's use it.
+    String startCursor = request.getParameter("cursor");
+    if (startCursor != null) {
+      fetchOptions.startCursor(Cursor.fromWebSafeString(startCursor));
+    }
+
+    // Query Datastore for all Comments, sorted by creation time
+    Query query = new Query("Comment").addSort("created", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:79]: I noticed that this is called in many places. I see that were calling it from a factory class, does the factory class enforce that this be a singleton?
https://en.wikipedia.org/wiki/Singleton_pattern

Usually we inject dependencies like this via the Guice framework or the equivalent.
This is not so important if the factory class is not in your control, but if it is consider making sure that you don't create the DatastoreService more than once.

**Comment by n-wach** [comment]
Oops, I pushed these to the wrong branch... I'm not sure if fola's reviews will stick around if I rebase, so I'll just expand this PR to be both adding GCP Datastore and adding pagination

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 1 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:32]: I was thinking there might be times where I want to create a comment "now", like when I just received a post request for a new comment.  But I can see how this constructor could make things unclear, especially since I never use it.  Will remove.

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:32]: Done

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 6 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:79]: The examples (for instance [here](https://github.com/n-wach/step-portfolio/blob/master/walkthroughs/week-3-server/examples/todo-list/src/main/java/com/google/sps/servlets/NewTaskServlet.java#L39)) call getDatastoreService() for every request, so I did the same.  I couldn't find anything specifically against/for calling it like this in the docs, but all the examples do this so I assume it's fine

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java (hunk 5 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:23]: `@AutoValue` looks really cool, I'll switch these over

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 4 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:34]: Fixed

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 3 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:19]: Fixed, along with `CommentThread`

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java (hunk 5 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java]: Done

**Comment by n-wach** [comment]
@f-camp Should I use `ImmutableList` in `CommentThread`?

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 2 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Done

**Comment by n-wach** [comment]
Oh and also, should I use a Builder instead of a static `create`?

**Comment by f-camp** [comment]
If possible yes please use ImmutableList.
Builder or Create either works. Create is ok if the number of params are not too many, if it turns out you have many params then a Builder is the best choice.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 6 above)
Comment (f-camp) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:79]: In general for resources like this it is best not to have to consistently instantiate this for every request.
How about if we have some private property
private DatastoreService datastore;
then override the [init](https://docs.oracle.com/javaee/7/api/javax/servlet/GenericServlet.html#init--) method for [HttpServlet](https://docs.oracle.com/javaee/7/api/javax/servlet/http/HttpServlet.html).

The init method is a good way to instantiate resources that would be used through out the lifecycle of a Servlet.

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    // Didn't work with these the other way around...
+    return new AutoValue_CommentThread(cursor, comments);
+  }
+
+  /** Returns a web-safe string that can be used to query for additional
+   * comments in this thread. This is null if there are no more comments. */
+  @Nullable public abstract String getCursor();
+
+  /** Returns the comment thread's comments. */
+  public abstract List<Comment> getComments();
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/CommentThread.java:37]: It is better to return Immutable if it shouldn't be changing.

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:37]: "immutable classes are easier to design, implement, and use than mutable classes. They are less prone to error and are more secure."

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+public class CommentsServlet extends HttpServlet {
+
+  static final int QUERY_SIZE = 5;
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the new comment from the form.
+    String comment = request.getParameter("comment");
+
+    // Create entity.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:53]: Comments should be explaining something the code cannot explain. Here, this code seems redundant.

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:53]: Many of the comments in this file do not seem necessary. Can you look over them.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    response.sendRedirect("/#comments");
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Limit the number of comments that can be loaded at once.
+    FetchOptions fetchOptions = FetchOptions.Builder.withLimit(QUERY_SIZE);
+
+    // If this servlet is passed a cursor parameter, let's use it.
+    String startCursor = request.getParameter("cursor");
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:72]: "cursor" seems like an important string. Strings which have a special meaning should be made into static final Strings.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    if (startCursor != null) {
+      fetchOptions.startCursor(Cursor.fromWebSafeString(startCursor));
+    }
+
+    // Query Datastore for all Comments, sorted by creation time
+    Query query = new Query("Comment").addSort("created", SortDirection.DESCENDING);
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:78]: typically you do not want to use raw strings, especially that they are used in multiple places. By using a static final String it makes it obvious to the user how the methods are related.

Another benefit is if you need to change the string, you only need to change it in one place instead of multiple places. If you forgot to change it in one of the instances, it would result in a bug.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    PreparedQuery pq = datastore.prepare(query);
+    QueryResultList<Entity> results;
+    try {
+      results = pq.asQueryResultList(fetchOptions);
+    } catch (IllegalArgumentException e) {
+      // IllegalArgumentException happens when an invalid cursor is used.
+      response.sendError(400, "Bad cursor");
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:87]: I would make the error message indicate an invalid cursor used. By doing this, you no longer need your comment.

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -35,27 +35,53 @@ const WELCOME_SELECTOR = "#welcome"
...
+    let p = document.createElement("p");
+    p.innerText = comment.content + " at " + comment.created;
+    element.appendChild(p);
+  }
+
+  if (thread.cursor) {
+    element.setAttribute("cursor", thread.cursor);
+    loadSource.disabled = false;
+  } else {
+    element.removeAttribute("cursor");
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:79]: Please make "cursor" into a const variable for reasons mentioned above.

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Contains a series of comments, with a cursor for pagination. */
+@AutoValue
+public abstract class CommentThread {
+  public static CommentThread create(List<Comment> comments, @Nullable String cursor) {
+    // Didn't work with these the other way around...
```
[the remark this answers is not in the document]
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:28]: I got a compilation error doing `new AutoValue_CommentThread(comments, cursor)`, it only worked with `cursor` first

[the remark this answers is not in the document]
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:28]: Switched the order of my getters fixed this

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    return new AutoValue_CommentThread(cursor, comments);
+  }
+
+  /** Returns a web-safe string that can be used to query for additional
+   * comments in this thread. This is null if there are no more comments. */
```
[the remark this answers is not in the document]
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:33]: Yeah I can use Optional here, it makes the JSON a little messier as `cursor` becomes an object with a `value` field, instead of simply a String

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java (hunk 7 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:37]: Changed to ImmutableList

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 11 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:87]: Smart, done

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 6 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Just tested and this seems to work.  Committed.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 8 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Removed superfluous comments

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 10 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 9 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Done

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js]: Done

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2019 Google LLC
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:1]:
```suggestion
// Copyright 2020 Google LLC
```

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2019 Google LLC
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/CommentThread.java:1]:
```suggestion
// Copyright 2020 Google LLC
```

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  public static CommentThread create(ImmutableList<Comment> comments) {
+    return new AutoValue_CommentThread(comments, Optional.empty());
+  }
+
+  /** Returns the comment thread's comments. */
+  public abstract ImmutableList<Comment> comments();
+
+  /** Returns a web-safe string that can be used to query for additional
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/CommentThread.java:36]: For multi line comments, they should have this form:

/**
 * bla
 */
https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+public abstract class CommentThread {
+  public static CommentThread create(ImmutableList<Comment> comments, Optional<String> cursor) {
+    return new AutoValue_CommentThread(comments, cursor);
+  }
+
+  public static CommentThread create(ImmutableList<Comment> comments) {
+    return new AutoValue_CommentThread(comments, Optional.empty());
+  }
+
+  /** Returns the comment thread's comments. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/CommentThread.java:33]: nit: when a method is obvious, the you can omit the comment. Here I think it is pretty obvious what this method does.

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+  public abstract ImmutableList<Comment> comments();
+
+  /** Returns a web-safe string that can be used to query for additional
+   * comments in this thread. This is null if there are no more comments. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/CommentThread.java:37]: This comment should be updated because it will never return null.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 20]
```diff
@@ -0,0 +1,130 @@
+// Copyright 2019 Google LLC
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:1]:
```suggestion
// Copyright 2020 Google LLC
```

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results) {
+      String content = (String) entity.getProperty(PROP_CONTENT);
+      Date created = (Date) entity.getProperty(PROP_CREATED);
+      comments.add(Comment.create(content, created));
+    }
+
+    CommentThread thread;
+    if (comments.size() == QUERY_SIZE) {
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:114]: To make it more obvious what is happening I would switch the if statement logic.

if (comments.size() < QUERY_SIZE) {
  // All comments were queried
  thread = CommentThread.create(ImmutatbleList.of(comments));
} else {
  // It is possible QUERY_SIZE limited the amount of comments obtained. Providing the cursor.
  thread = CommentThread. create(ImmutableList.copyOf(comments), Optional.of(cursorString));
}

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results) {
+      String content = (String) entity.getProperty(PROP_CONTENT);
+      Date created = (Date) entity.getProperty(PROP_CREATED);
+      comments.add(Comment.create(content, created));
+    }
+
+    CommentThread thread;
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:113]: mark this final so it is made clear that this will be instantiated.

File: portfolio/src/main/webapp/script.js [hunk 23]
```diff
@@ -22,40 +22,68 @@ const PROJECT = "project";
...
-for(let project of document.getElementsByClassName(PROJECT)) {
+for (let project of document.getElementsByClassName(PROJECT)) {
   project.addEventListener("click", openProject);
 }
 
 
 /**
- * Fetches and displays a welcome message from DataServlet (at /data).
+ * Fetches and displays a thread from CommentsServlet (at /comments).
+ * Adds comments as children to the provided DOM element. This parent element can have a
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:55]: For new paragraphs you want to add \<p\>

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+  private static final int QUERY_SIZE = 5;
+
+  private static final String PARAM_CONTENT = "comment";
+  private static final String PARAM_CURSOR = "cursor";
+
+  private static final String KIND_COMMENT = "Comment";
+
+  private static final String PROP_CONTENT = "content";
+  private static final String PROP_CREATED = "created";
+
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:59]: No spaces between params of same declaration.

In class organization we should follow a consistent flow so that other engineers can easily get used to our code base.
I would suggest a flow like:
public static final ....
public static final ...

private static final ...
private static final...

private ...
private ...

Is extremely common.
https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String comment = request.getParameter(PARAM_CONTENT);
+
+    Entity commentEntity = new Entity(KIND_COMMENT);
+    commentEntity.setProperty(PROP_CONTENT, comment);
+    commentEntity.setProperty(PROP_CREATED, new Date());
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:75]: reuse the initialized datastore from init?

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+      response.sendError(400, "Invalid cursor supplied");
+      return;
+    }
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results) {
+      String content = (String) entity.getProperty(PROP_CONTENT);
+      Date created = (Date) entity.getProperty(PROP_CREATED);
+      comments.add(Comment.create(content, created));
+    }
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:111]:
```suggestion
  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
    FetchOptions fetchOptions = FetchOptions.Builder.withLimit(QUERY_SIZE);
    String startCursor = request.getParameter(PARAM_CURSOR);
    if (startCursor != null) {
      fetchOptions.startCursor(Cursor.fromWebSafeString(startCursor));
    }

    Query query = new Query(KIND_COMMENT).addSort(PROP_CREATED, SortDirection.DESCENDING);
    PreparedQuery pq = datastore.prepare(query);
    QueryResultList<Entity> results;
    try {
      results = pq.asQueryResultList(fetchOptions);
    } catch (IllegalArgumentException e) {
      response.sendError(400, "Invalid cursor supplied");
      return;
    }

    List<Comment> comments = new ArrayList<>();
    for (Entity entity : results) {
      String content = (String) entity.getProperty(PROP_CONTENT);
      Date created = (Date) entity.getProperty(PROP_CREATED);
      comments.add(Comment.create(content, created));
    }
```

No need for comment when your code is self explanatory. Good code always tells the story which you're doing :)

File: portfolio/src/main/webapp/script.js [hunk 27]
```diff
@@ -22,40 +22,68 @@ const PROJECT = "project";
...
+    element.setAttribute(ATTR_CURSOR, thread.cursor.value);
+    loadSource.disabled = false;
+  } else {
+    element.removeAttribute(ATTR_CURSOR);
+    loadSource.remove();
+  }
 }
 
-showWelcome();
+// Load some of the comments.
```
Comment (f-camp) [inline portfolio/src/main/webapp/script.js:86]: nit: this can be removed as the code is self explanatory.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+
+    CommentThread thread;
+    if (comments.size() == QUERY_SIZE) {
+      // More comments can (probably) be loaded from here.
+      String cursorString = results.getCursor().toWebSafeString();
+      thread = CommentThread.create(ImmutableList.copyOf(comments), Optional.of(cursorString));
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:117]: Can use comments.build() if you use an ImmutableList.Builder to build the comments list here.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+    } else {
+      thread = CommentThread.create(ImmutableList.copyOf(comments));
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:119]: ditto.

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 15 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:1]: Fixed

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java (hunk 16 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:1]: Fixed

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java (hunk 18 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:33]: Fixed

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java (hunk 19 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:37]: Changed "null" to "empty"

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 20 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:1]: Fixed

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 21 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:114]: Done, good idea

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 22 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:113]: Done

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 24 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:59]: Fixed, thanks

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 26 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:111]: Sure.  Guess it just takes time to learn what's a good amount of comments :P

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 29 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:119]: Done

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 28 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:117]: Didn't know this was a thing, done

File: portfolio/src/main/webapp/script.js (hunk 27 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:86]: I'm gonna keep it cause the .onclick is defined in another file and it might be a little unclear what it does...

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 25 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:75]: Whoops, done

File: portfolio/src/main/webapp/script.js (hunk 23 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:55]: JSDoc doesn't use <p>, but I added an empty line as recommended by the styleguide

Comment (n-wach) [reply portfolio/src/main/webapp/script.js:55]: `<p>`

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+
+  public static CommentThread create(ImmutableList<Comment> comments) {
+    return new AutoValue_CommentThread(comments, Optional.empty());
+  }
+
+  public abstract ImmutableList<Comment> comments();
+
+  /**
+   * Returns a web-safe string that can be used to query for additional
+   * comments in this thread. This empty if there are no more comments.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/CommentThread.java:37]:
```suggestion
   * comments in this thread. If there are no more comments to query, returns {@code Optional.empty}.
```

File: portfolio/src/main/webapp/script.js [hunk 31]
```diff
@@ -16,46 +16,74 @@
...
+const COMMENTS_LOADER = "#comments-loader"
 
+const ATTR_CURSOR = "cursor";
+const PARAM_CURSOR = "cursor";
 
 /**
  * Handles click event for project div.
- * 
+ *
  * When a project div is clicked and currently closed, this expands and scrolls
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:33]: Use of "this" and "it" add some slight confusion as to what is being reference. Consider rewriting it as:

When the project div is closed, then clicking it results in the div being expanded and scrolled into view. Otherwise, the div is shrunk back into its default size.

File: portfolio/src/main/webapp/script.js [hunk 32]
```diff
@@ -16,46 +16,74 @@
...
 // and ready.
-for(let project of document.getElementsByClassName(PROJECT)) {
+for (let project of document.getElementsByClassName(PROJECT)) {
   project.addEventListener("click", openProject);
 }
 
 
 /**
- * Fetches and displays a welcome message from DataServlet (at /data).
+ * Fetches and displays a thread from CommentsServlet (at /comments).
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:53]: Can you used {@code CommentsServlet} ?

File: portfolio/src/main/webapp/script.js [hunk 33]
```diff
@@ -16,46 +16,74 @@
...
+ *
+ * Adds comments as children to the provided DOM element. This element can have a
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:55]: I would rephrase this as:

Adds comments as children to the provided DOM {@code element}. If the element contains a cursor, then all of its remaining comments are loaded < or mention how many comments are loaded>. Removes the {@code loadSource} if there are no more comments left to load.

File: portfolio/src/main/java/com/google/sps/data/CommentThread.java (hunk 30 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/CommentThread.java:37]: Done

File: portfolio/src/main/webapp/script.js (hunk 31 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:33]: Done

File: portfolio/src/main/webapp/script.js (hunk 32 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:53]: Done

File: portfolio/src/main/webapp/script.js (hunk 33 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js]: Reworded

**Comment by n-wach** [comment]
@f-camp merging this to unblock, I'll ping you Monday in case you want to review it
